1. **Prueba que si** $p$ **es primo impar,** $2(p-3)! \equiv -1 \text{ mod } p$ 



Usaremos el Teorema de Wilson, que afirma que dado $p$ primo se cumple que $(p-1)! \equiv -1 \text{ mod }p$ 

Tengamos en cuenta que $(p-1) \equiv_p -1$, $(p-2) \equiv_p -2$ y que $(p-1)! = (p-1)(p-2)(p-3)!$ . Con esta información, obtenemos:

$-1 \equiv_p (p-1)! \equiv_p (p-1)(p-2)(p-3)! \equiv_p (-1)(-2)(p-3)! \equiv_p 2(p-3)!$ 

Por tanto, podemos concluir que $2(p-3)! \equiv -1 \text{ mod } p$ 


**2. Muestra que \-1 es residuo cuadrático para los primos de la forma 4k\+1y no lo es para los de la forma 4k\+3.**


Por el lema de Euler sabemos que $(\frac{a}{p})=a^{\frac{p-1}{2}} \ mod \ p$ . Sea entonces $a=-1$ y $p=4k+1$ tenemos:
\begin{equation*}
\left(\frac{-1}{4k+1}\right) =(-1)^{\frac{4k+1-1}{2}}=(-1)^{\frac{4k}{2}}=(-1)^{2k}=1
\end{equation*}
Por tanto, $-1$ si es residuo cuadrático para los primos de la forma $4k+1$.<br>
Volvemos a usar el lema de Euler para los primos de la forma $4k+3$ y obtenemos lo siguiente:
\begin{equation*}
\left(\frac{-1}{4k+3}\right) =(-1)^{\frac{4k+3-1}{2}}=(-1)^{\frac{4k+2}{2}}=(-1)^{2k+1}=-1
\end{equation*}
Por lo tanto, $-1$ no es residuo cuadrático para los primos de la forma $4k+3$.



3. **Prueba que si** $a$ **es residuo cuadrático de** $p$**, entonces** $a^{\frac{p+1}{4}}$ **es una raíz cuadrada de** $a$ **mod** $p$ **cuando** $p ≡ 3$ **mod** $4$**.**



Basta ver que $(a^{\frac{p+1}{4}})^2 \equiv a$ mod $p$.

Notamos que como $p \equiv 3$ mod $4 \Longrightarrow p+1$ es divisible por $4$.

Como $a$ es residuo cuadrático, entonces por Euler $a^{\frac{p-1}{2}} \equiv  \left( \frac{a}{p} \right) \equiv 1$ mod $p$. Entonces:

$(a^{\frac{p+1}{4}})^2 \equiv a^{\frac{p+1}{2}} \equiv a^{\frac{p-1}{2}}\cdot a \equiv 1\cdot a \equiv a$ mod $p$

4. **Encuentra los residuos cuadráticos** $\text{ mod } 29$ 



Los residuos cuadráticos del primo $p$  son $1, 2^2, 3^2, \dots, \left(\dfrac{p-1}{2}\right)^2$ . Los calcularemos con python y teniendo en cuenta que $\frac{p-1}{2} = 14$  



In [3]:
def modulo(numero, modulo):
    resto = numero//modulo
    numero_mod = numero - resto*modulo
    return numero_mod

residuos = []
for i in range(1, 15):
    residuos.append(modulo(i**2, 29))
residuos.sort()
    
print('Los residuos cuadráticos módulo 29 son: ', residuos)

Los residuos cuadráticos módulo 29 son:  [1, 4, 5, 6, 7, 9, 13, 16, 20, 22, 23, 24, 25, 28]


**5. Calcula el resto de la división de $3\cdot 6 \cdot 9 \cdot \cdot \cdot 306$ por 103.**

Tenemos que $3\cdot 6 \cdot 9 \cdot \cdot \cdot 306 = 3^{102}(1\cdot 2 \cdot 3 \cdot \cdot \cdot 102)= 3^{102} (102!)$. Vamos a calcular por separado $3^{102} $ y $(102)!$<br>
Entonces $\varphi(103)=102$ por ser 103 primo, por tanto como 3 y 103 son coprimos tenemos que $3^{\varphi(103)}=3^{102} =1 \mod 103 $ por el Teorema de Euler.<br>
Ahora por el Teorema de Wilson nos queda que $(103-1)! \equiv 102! \equiv -1 \ mod \ 103 \equiv 102 \ mod \ 103$. <br>
Entonces, el resto de la división de $3\cdot 6 \cdot 9 \cdot \cdot \cdot 306$ por 103 es
\begin{equation*}
3\cdot 6 \cdot 9 \cdot \cdot \cdot 306 \ mod \ 103 \equiv 3^{102} (102!) \equiv 1 \cdot 102 \equiv 102 \ mod \ 103
\end{equation*}
Es decir, que el resto de la divisón de $3\cdot 6 \cdot 9 \cdot \cdot \cdot 306$ por 103 es 102.

6. **Calcula** $\left( \frac{3}{11} \right)$**,** $\left( \frac{7}{11} \right)$**, y** $\left( \frac{19}{101} \right)$**.**



Para el primero, observamos que $11 \equiv 3$ mod $4$:

$\left( \frac{3}{11} \right) = -\left( \frac{11}{3} \right) = -\left( \frac{2}{3} \right) \equiv -2^{\frac{3-1}{2}}$ mod $3 \equiv -2$ mod $3 \equiv 1$

Para el segundo tenemos en cuenta que $7 \equiv 11 \equiv 3$ mod $4$:

$\left( \frac{7}{11} \right) = -\left( \frac{11}{7} \right) = -\left( \frac{4}{7} \right) = -\left( \frac{2}{7} \right)^2 = -1$

Para el tercero, observamos que $19 \equiv 3$ mod $4$:

$\left( \frac{19}{101} \right) = (-1)^{\frac{101-1}{2} \cdot \frac{19-1}{2}} \cdot \left( \frac{101}{19} \right) = (-1)^{450}\cdot \left( \frac{101}{19} \right) = \left( \frac{101}{19} \right) = \left( \frac{6}{19} \right) = \left( \frac{2}{19} \right) \cdot \left( \frac{3}{19} \right) \equiv 2^{\frac{19-1}{2}} \cdot \left( \frac{3}{19} \right)$ mod $19 \equiv  512 \cdot \left( \frac{3}{19} \right)$ mod $19 \equiv (-1) \cdot \left( \frac{3}{19} \right)$ mod $19 = - \left(- \left( \frac{19}{3} \right) \right) = \left( \frac{1}{3} \right) = 1$


7. **Encuentra todos los primos impares para los que** $3$ **es residuo cuadrático.**



Usaremos el símbolo de Legendre como ayuda para realizar el ejercicio. Tenemos que

$\left( \dfrac{3}{p} \right) = \begin{cases} 1 \mbox{ si } 3 \mbox{ es RC de   } p \\ -1 \mbox{ si } 3 \mbox{ es no RC de   } p \end{cases}$ 

Por la ley de reciprocidad cuadrática, como $p$ es primo impar, se tiene que 

$\left( \dfrac{3}{p} \right) = \begin{cases} \left( \dfrac{p}{3} \right) \mbox{ si } p \equiv 1 \mod 4 \\ -\left( \dfrac{p}{3} \right) \mbox{ si } p \equiv 3 \mod 4 \end{cases}$ 

Contemplamos ahora los casos de $p$ en módulo 3:

Si $p \equiv 1 \mod 3$ , entonces $\left( \dfrac{p}{3} \right) = \left( \dfrac{1}{3} \right) = 1$ .

Si $p \equiv 2 \mod 3$ , entonces $\left( \dfrac{p}{3} \right) = \left( \dfrac{2}{3} \right) = \left( \dfrac{-1}{3} \right) = (-1)^{\frac{3-1}{2}} =  -1$ .

Por tanto, llegamos a que 3 es residuo cuadrático si:

- O bien $p \equiv 1 \mod 4$  y $p \equiv 1 \mod 3$
- O bien  $p \equiv 3 \mod 4$  y $p \equiv 2 \mod 3$ 



**8. Encuentra todas las soluciones de la congruencia $x^2 \equiv 2 \ mod \ 47$.**

Veamos primero si dicha ecuación tiene solución (si lo hace tendrá dos ya que $2 \neq 0$). Para ello tenemos que ver si 2 es residuo cuadrático de 47 y en clase hemos visto un Teorema que nos dice que 2 es residuo cuadrático de un primo si este es de la forma $8k+1$ o $8k-1$, en este caso $47=8 \cdot 6 -1$, por lo tanto 2 es residuo cuadrático de 47 y por tanto dicha ecuación tendrá dos soluciones.<br>
Ahora ya podemos calcularlas. Para ello observamos que $47 \equiv 3 \ mod \ 4$ entonces, por un teorema visto en clase, la raíz cuadrada es $2^{\frac{47+1}{4}} \equiv 2^{12} \equiv 4096 \equiv 7 \ mod \ 47$. Por tanto 7 y -7, es decir, 7 y 40 son las raíces de 2 módulo 47.<br>
Lo comprobamos y vemos que $7^2 \equiv 49\equiv 2 \ mod \ 47$ y $40^2 \equiv 1600 \equiv 2 \ mod \ 47$.

9. **Sean** $p$**,** $q$ **primos impares y** $n = pq$**. Suponed que** $gcd(a, n) = 1$**. Si existe una solución para** $x^2 = a$ **mod** $n$**, probad que tiene cuatro soluciones.**



Dado que $(a,n) = 1$ y $n=pq$, entonces $p \nmid a$ y $q \nmid a$.

Por el teorema Chino de los restos, como $(p,q)=1$ por ser $p$ y $q$ primos, que exista una solución de $x^2 \equiv a$ mod $n$ es equivalente a que exista solución del sistema:

\begin{equation*}
x^2 \equiv a \ mod \ p
\end{equation*}

\begin{equation*}
x^2 \equiv a \ mod \ q
\end{equation*}

Así, como sabemos que $x^2 \equiv a$ mod $n$ tiene una solución, entonces $x^2 \equiv a$ mod $p$ tendrá solución y $x^2 \equiv a$ mod $q$ también. Es decir, que $a$ es residuo cuadrático de $p$ y de $q$.

Como $a$ es residuo cuadrático de $p$ entonces $x^2 \equiv a$ mod $p$ tiene exactamente dos soluciones distintas, digamos $x_1$ y $x_2$.

Análogamente, como $a$ es residuo cuadrático de $q$ entonces $x^2 \equiv a$ mod $q$ tiene exactamente dos soluciones distintas, digamos $y_1$ e $y_2$.

De hecho, estas soluciones cumplirán que $x_2 \equiv -x_1$ mod $p$ y que $y_2 \equiv -y_1$ mod $q$, y serán las raíces cuadradas de $a$ mod $p$ y mod $q$ respectivamente.

Ahora, combinando estas soluciones y aplicando el teorema Chino de los restos, podemos resolver 4 sistemas para obtener las soluciones de $x^2 \equiv a$ mod $n$.

La primera solución la obtendríamos aplicando el teorema Chino de los restos al siguiente sistema:

\begin{equation*}
x \equiv x_1 \ mod \ p
\end{equation*}

\begin{equation*}
x \equiv y_1 \ mod \ q
\end{equation*}

Para obtener la segunda resolvemos el sistema:


\begin{equation*}
x \equiv x_1 \ mod \ p
\end{equation*}

\begin{equation*}
x \equiv y_2 \ mod \ q
\end{equation*}

Para la tercera resolvemos:


\begin{equation*}
x \equiv x_2 \ mod \ p
\end{equation*}

\begin{equation*}
x \equiv y_1 \ mod \ q
\end{equation*}

Y por último para la cuarta solución resolvemos:


\begin{equation*}
x \equiv x_2 \ mod \ p
\end{equation*}

\begin{equation*}
x \equiv y_2 \ mod \ q
\end{equation*}

Así, el teorema Chino de los restos nos daría una solución única módulo $pq = n$ para cada sistema, que será solución de $x^2 \equiv a$ mod $n$, de forma que en total obtendríamos 4 soluciones de $x^2 \equiv a$ mod $n$.

10. **Sean** $p, q$ **primos impares y** $n = pq$**. Sean** $b,c$ **enteros. ¿Cuáles son las posibles soluciones** $x \in \mathbb{Z}_n$  **de** $x^2 +bx +c = 0 \mod n$ **?**



Hacemos mínimos cuadrados en la expresión con el objetivo de encontrar una ecuación de la forma $x^2  \equiv a \mod n$ . Notemos que podemos usar el término $2^{-1} \mod n$ ya que se tiene que $p,q$ son primos impares y por tanto $(2,n) = 1$. Obtenemos pues:

$x^2 + bx + c \equiv_n 0 \Leftrightarrow x^2 + bx + \left(2^{-1}b\right)^2 - \left(2^{-1}b\right)^2 + c \equiv_n 0 \Leftrightarrow  \left(x+ 2^{-1}b\right)^2  \equiv_n \left(2^{-1}b\right)^2 - c$ 

Por definición, esta ecuación tendrá solución si $\left(2^{-1}b\right)^2 - c$ es RC de $p \text{ y } q$, ya que $n = pq$. 

Del ejercicio anterior, podemos aplicarlo y llamar $a_1, a_2, a_3, a_4$ a las posibles soluciones de la ecuación $x^2 \equiv_n (2^{-1}b)^2 - c$ .

Por tanto, en este caso las posibles soluciones tendrían la forma  $a_i - (2^{-1}b) \text{ para } i = 1,2,3,4.$ 


**11. Calcula $\left( \frac{15}{17} \right)$. Tiene soluciones $7x^2+x+11 \equiv 0 \ mod \ 17$?**

Primero calculamos $\left( \frac{15}{17} \right)$ y nos queda
\begin{equation*}
\left( \frac{15}{17} \right) =\left( \frac{3}{17} \right) \left( \frac{5}{17} \right) =(*) \left( \frac{17}{3} \right) \left( \frac{17}{5} \right)=\left( \frac{2}{3} \right) \left( \frac{2}{5} \right)=(**) (-1)(-1)=1
\end{equation*}
$(*)$ es porque $17 \equiv \ 1 \ mod \ 4$ entonces no hace falta cambiar ningún signo y $(**)$ es porque solo la mitad de p-1 son residuos cuadráticos, en módulo 3 solo tendremos un residuo cuadrático y este es el 1 ya que $1^2 \equiv 1 \mod \ 3$ por tanto 2 no lo puede ser y en módulo 5 solo tendremos 2 residuos cuadráticos y lo son el 1 y el 4 porque $1^2 \equiv 1 \mod \ 5$ y $2^2 \equiv 4 \mod \ 5$, es decir, que los dos números de Legendre son -1. Por lo tanto$\left( \frac{15}{17} \right)=1$.<br>
Vemos ahora la ecuación $7x^2+x+11 \equiv 0 \ mod \ 17$, primero multiplicamos toda la ecuación por $7^{-1} \equiv 5 \ mod \ 17 $ y nos queda la siguiente ecuación
\begin{equation*}
7x^2+x+11 \equiv 0 \ mod \ 17 \Leftrightarrow 5 \cdot 7x^2+5 \cdot x+5 \cdot 11 \equiv 5 \cdot 0 \ mod \ 17 \Leftrightarrow x^2+5x+55 \equiv 0 \ mod \ 17 \equiv 5 \cdot 0 \ mod \ 17 \Leftrightarrow x^2+5x+4 \equiv 0 \ mod \ 17
\end{equation*}
Ahora hacemos mínimos cuadrados y nos queda:
\begin{equation*}
(x+2^{-1}5)^2 \equiv (*)(x+9 \cdot 5)^2\equiv (x+45)^2 \equiv (x+11)^2
\end{equation*}
$(*)$ es porque el inverso de 2 módulo 17 es 9. De aquí nos queda que $(x+11)^2-11^2+4 \equiv 0 \ mod \ 17 \Leftrightarrow (x+11)^2 -2+4 \equiv 0 \ mod \ 17 \Leftrightarrow (x+11)^2 +2 \equiv 0 \ mod \ 17 \Leftrightarrow (x+11)^2 \equiv -2 \ mod \ 17  \Leftrightarrow (x+11)^2 \equiv 15 \ mod \ 17$. Por tanto, la ecuación tendrá solución si 15 es residuo cuadrático de 17, lo cual ya lo hemos probado al inicio al ver que $\left( \frac{15}{17} \right)=1$. Por tanto, la ecuación $7x^2 +x+11 \equiv 0 \ mod \ 17$ tiene dos soluciones (por ser 15 distinto de 0).

12. **Calcula**  $\left( \frac{17}{19} \right)$**. ¿ Tiene soluciones** $5x^2 + 4x +7 \equiv 0$ **mod** $19$**?**



$\left( \frac{17}{19} \right) = (-1)^{\frac{19-1}{2} \cdot \frac{17-1}{2}} \cdot \left( \frac{19}{17} \right) = (-1)^{72}\cdot \left( \frac{19}{17} \right) = \left( \frac{19}{17} \right) = \left( \frac{2}{17} \right) \equiv 2^{\frac{17-1}{2}}$ mod $17 \equiv 2^8$ mod $17 \equiv 256$ mod $17 \equiv 1$

Transformamos $5x^2 + 4x +7 \equiv 0$ mod $19$ para ver si tiene solución. En primer lugar, multiplicamos la ecuación por $5^{-1} \equiv 4$ mod $19$, y obtenemos:

$x^2 + 16x +9 \equiv 0$ mod $19$

Para completar cuadrados, tenemos en cuenta que:

$2^{-1} \equiv 10$ mod $19 \Longrightarrow 16 \cdot 2^{-1} \equiv 16 \cdot 10 \equiv 8$ mod $19$

Así, obtenemos que la anterior ecuación es equivalente a:

$(x+8)^2 - 8^2 + 9 \equiv 0$ mod $19 \Longleftrightarrow (x+8)^2 \equiv 17$ mod $19$

De esta forma, como $ \left( \frac{17}{19} \right) = 1$, tenemos que $(x+8)^2 \equiv 17$ mod $19$ sí tiene solución o, lo que es lo mismo, que $5x^2 + 4x +7 \equiv 0$ mod $19$ sí tiene solución.

13. **Considera el número** $b = 34666547679800897875547457475563635683$. **Muestra que** $53$ **tiene raíces** $\mod b$ **y encuéntralas.**



Para ver que $53$ tiene raíces mod $b$, veamos que es residuo cuadrático usando el número de Legendre. Notemos que $53 \equiv 1 \mod 4$ y que $b \equiv 47 \mod 53$, por lo que:

$\left( \dfrac{53}{b} \right) = \left( \dfrac{b}{53} \right) = \left( \dfrac{47}{53} \right) = \left( \dfrac{53}{47} \right) = \left( \dfrac{6}{47} \right) = \left( \dfrac{2}{47} \right) \left( \dfrac{3}{47} \right)$ 

Como el $47$ es primo impar, usamos que $\left( \dfrac{2}{47} \right) \equiv_{47} 2^{\frac{46}{2}} \equiv_{47} 1$ y que $\left( \dfrac{3}{47} \right) \equiv_{47} 3^{\frac{46}{2}} \equiv_{47} 1$ , por lo que llegamos a que $\left( \dfrac{53}{b} \right) = 1$ y 53 es RC de $b$. 

Para calcular las raíces, usaremos el hecho de que $b \equiv 3\mod 4$ , por lo que un teorema de clase nos da las raíces de manera directa, siendo estas $53^{\frac{b+1}{4}} \text{ y } -53^{\frac{b+1}{4}}$ 

Ejecutamos en python los comandos necesarios para obetenerlas:



In [3]:
b = 34666547679800897875547457475563635683
sol1 = pow(53, (b+1)//4, b)
sol2 = pow(-sol1,1,b)
print('Las dos raíces módulo b son', sol1, 'y', sol2)

Las dos raíces módulo b son 10612286401713577499344834229287288295 y 24054261278087320376202623246276347388


**14. Construye un criptosistema de Rabin para cifrar el mensaje OR. ¿qué ventajas y desventajas tiene el sistema de Rabin respecto del RSA? ¿Es más seguro?**

Tomamos un alfabeto de 27 letras donde el 0 corresponde al espacio. Tomamos los primos p=31 y q=43 (los dos congruentes con 3 módulo 4) y por tanto nuestra n será $n=pq=31 \cdot 43 =1333$.<br>
Nuestro mensaje a cifrar es OR que si lo pasamos a número obtenemos $15 +18 \cdot 27 = 501 = m$. Ahora elevamos dicho mensaje al cuadrado para cifrarlo y nos queda:
\begin{equation*}
c\equiv m^2 \equiv 501^2 \equiv 251001 \equiv 397 \ mod \ 1333
\end{equation*}
y ahora pasamos a letras el mensaje cifrado y obtenemos $397=19 + 14 \cdot 27$ y como el 19 corresponde a la S y el 14 corresponde a la N el mensaje cifrado será "SN".<br>

Ahora procedemos a descifrar dicho mensaje, al hacerlo obtendremos 4 soluciones y solo una de ellas será nuestro mensaje inicial.

Primero calculamos las raíces módulo p y módulo q, que como son congruentes con 3 módulo 4 es sencillo:
\begin{equation*}
\left\{ m_p \equiv c^{\frac{p+1}{4}} \equiv 397^{\frac{31+1}{4}} \equiv 397^8 \equiv 5 \ mod \ 31   \atop m_q \equiv c^{\frac{q+1}{4}} \equiv 397^{\frac{43+1}{4}} \equiv 397^{11} \equiv 15 \ mod \ 43 \right.
 \end{equation*}
Por tanto, las posibles soluciones serán:
\begin{equation*}
\left\{ x= \pm 5 \ mod \ 31 \atop x= \pm 15 \ mod \ 43 \right.
 \end{equation*}
Vamos a calcular  los cuatros posibles mensajes y uno de ellos deberá ser 501 que es nuestro mensaje inicial. Para ello utilizamos el Teorema Chino de los restos. Empezamos por el caso:
\begin{equation*}
\left\{ x= 5 \ mod \ 31 \atop x= 15 \ mod \ 43 \right.
 \end{equation*}
Nuestra solución vendrá dada por $N=5 \cdot 43 \cdot (43)^{-1}_{31} + 15 \cdot 31 \cdot (31)^{-1}_{43}$. Los inversos nos dan 13 y 25 respectivamente por lo que la solución en este caso es $N \equiv 5 \cdot 43 \cdot 13 + 15 \cdot 31 \cdot 24 \equiv 14420 \equiv 1090 \ mod \ 1333$ que no es nuestro mensaje inicial. Lo intentamos ahora con 
\begin{equation*}
\left\{ x= 5 \ mod \ 31 \atop x= -15 \ mod \ 43 \right.
 \end{equation*}
y procediendo igual que antes tenemos $N\equiv 5 \cdot 43 \cdot 13 - 15 \cdot 31 \cdot 25 \equiv -8830 \equiv 501 \ mod \ 1333$. En este caso si que es nuestro mensaje inicial. Igualmente vamos a calcular los dos casos que faltan porque en principio no sabríamos cuál es el real. Hacemos ahora 
\begin{equation*}
\left\{ x= -5 \ mod \ 31 \atop x= 15 \ mod \ 43 \right.
 \end{equation*}
y procediendo de la misma forma que antes tenemos $N\equiv -5 \cdot 43 \cdot 13 + 15 \cdot 31 \cdot 25 \equiv 8830 \equiv 832 \ mod \ 1333$. Ahora el último caso que es
\begin{equation*}
\left\{ x= -5 \ mod \ 31 \atop x= -15 \ mod \ 43 \right.
 \end{equation*}
Cuya solución es $N\equiv -5 \cdot 43 \cdot 13 - 15 \cdot 31 \cdot 25 \equiv -14420 \equiv 243 \ mod \ 1333$.<br>
Obtenemos así las cuatro soluciones posibles al descrifrar y comprobamos que una de ellas es nuestro mensaje inicial.<br>
Es más seguro el sistema de Rabin que el RSA ya que hemos visto un Teorema en clase que nos dice que la complejidad de quebrar Rabin es equivalente a la complejidad de factorizar n, mientras que el RSA se puede quebrar sin que esto implique que se pueda factoizar n.<br>
El sistema de Rabin tiene la ventaja respecto al RSA de que encriptar y desencriptar (si p y q son congruentes a 3 módulo 4) es mucho más rápido, sin embargo, tenemos la desventaja de que al desencriptar obtenemos 4 posibles mensajes sin saber, en principio, cual es el original.


15. **Encuentra primos** $p$**,** $q$ **tal que** $n = pq$ **permite cifrar mensajes de 5 en 5 letras usando el criptosistema de Rabin. Cifra GAUSS usando** $n$ **y verifica que cuando descifras, recuperas el mensaje.**



Tomamos un alfabeto con 27 letras.

Queremos $p$ y $q$ primos tales que $p \equiv q \equiv 3$ mod $4$, y además, para poder cifrar mensajes de 5 en 5 letras, debemos tomar $n=pq> 27^5 = 14348907$.

Tomamos por ejemplo $p= 3803$ y $q=3823$ que cumplen ambas condiciones:

In [4]:
p = 3803
q = 3823

n = p*q
n

14538869

In [5]:
n>27**5

True

In [6]:
p%4, q%4

(3, 3)

Para cifrar GAUSS se hace lo siguiente:

$m =$ GAUSS $= (7,1,21,19,19)_{27} = 7 \cdot 27^4 + 1 \cdot 27^3 + 21 \cdot 27^2 + 19 \cdot 27^1 + 19 \cdot 27^0 = 3755611$

Ahora, Bob para encriptar $m$ calcula:

$c \equiv m^2$ mod $n \equiv 3755611^2$ mod $14538869 \equiv 6461482 = (12,4,7,13,4)_{27} =$ LDGMD

y se lo envía a Alice.

In [7]:
m = 7*27**4 + 1*27**3 + 21*27**2 + 19*27**1 + 19*27**0
c = (m**2)%n
m, c

(3755611, 6461482)

Para descifrar, Alice hace las raíces cuadradas de $c$ mod $p$ y mod $q$:

$m_p \equiv c^{\frac{p+1}{4}}$ mod $p \equiv 6461482^{\frac{3803+1}{4}}$ mod $3803 \equiv 1753$

$m_q \equiv c^{\frac{q+1}{4}}$ mod $q \equiv 6461482^{\frac{3823+1}{4}}$ mod $3823 \equiv 1425$

Usando el teorema Chino de los restos con $(m_p,m_q)$, $(-m_p,m_q)$, $(m_p,-m_q)$ y $(-m_p,-m_q)$ se obtienen 4 posibles mensajes resolviendo cada uno de los siguientes sistemas:

\begin{equation*}
x \equiv 1753 \ mod \ 3803
\end{equation*}

\begin{equation*}
x \equiv 1425 \ mod \ 3823
\end{equation*}

$M = 3803 \cdot 3823 = 14538869$

$M_1 = \frac{14538869}{3803} = 3823$

$M_2 = \frac{14538869}{3823} = 3803$

$y_1 = 3823^{-1}$ mod $3803 \equiv 2472$

$y_2 = 3803^{-1}$ mod $3823 \equiv 1338$

$x \equiv 1753 \cdot 3823 \cdot 2472 + 1425 \cdot 3803 \cdot 1338$ mod $14538869 \equiv 2971896 = (5,15,26,18,6)_{27} =$ EOZRF

\begin{equation*}
x \equiv -1753 \ mod \ 3803
\end{equation*}

\begin{equation*}
x \equiv 1425 \ mod \ 3823
\end{equation*}

$x \equiv -1753 \cdot 3823 \cdot 2472 + 1425 \cdot 3803 \cdot 1338$ mod $14538869 \equiv 3755611 = (7,1,21,19,19)_{27} =$ GAUSS

\begin{equation*}
x \equiv 1753 \ mod \ 3803
\end{equation*}

\begin{equation*}
x \equiv -1425 \ mod \ 3823
\end{equation*}

$x \equiv 1753 \cdot 3823 \cdot 2472 - 1425 \cdot 3803 \cdot 1338$ mod $14538869 \equiv 10783258 = (20,7,22,22,25)_{27} =$ TGVVY

\begin{equation*}
x \equiv -1753 \ mod \ 3803
\end{equation*}

\begin{equation*}
x \equiv -1425 \ mod \ 3823
\end{equation*}

$x \equiv -1753 \cdot 3823 \cdot 2472 - 1425 \cdot 3803 \cdot 1338$ mod $14538869 \equiv 11566973 = (21,20,17,2411)_{27} =$ UTQXK

Los 4 posibles mensajes son EOZRF, GAUSS, TGVVY y UTQXK, por lo que Alice elige GAUSS.

In [8]:
mp = 1753
mq = 1425

def euclides(a, b):
    
    if b == 0:
        return (a, 1, 0)
    
    else:
        q = a // b
        r = a % b
        (mcd, u0, v0) = euclides(b, r)
        u = v0
        v = u0-q*v0
        return (mcd, u, v)

def inversa(a,m):
    v = euclides(a,m)[1]
    return v % m

y1 = inversa(q,p)
y2 = inversa(p,q)

x1 = (mp*q*y1 + mq*p*y2) % n
x2 = (-mp*q*y1 + mq*p*y2) % n
x3 = (mp*q*y1 - mq*p*y2) % n
x4 = (-mp*q*y1 - mq*p*y2) % n

x1, x2, x3, x4

(2971896, 3755611, 10783258, 11566973)

16. **Considera el protocolo de Diffie\-Hellman con** $q=3$ **y** $p=353$. **Alice elige** $a=97$ **y** **Bob elige** $b=233$. **Calcula** $A$ **y** $B$ **y la clave. Diseña una extensión del protocolo para tres personas \(Alice, Bob y Carlos\) para que se intercambien una clave secreta común.**



Consideramos $g=3$ la raíz primitiva de $p = 233$. Por los procedimientos vistos en clase, tenemos que:

- $A = g^a \mod p \equiv 3^97 \mod p \equiv 40 \mod p$ 
- $B = g^b \mod p \equiv 3^{233} \mod p \equiv 248 \mod p$



La clave en este caso viene dada por $g^{ab} \equiv_p 160$. 


Para extender el protocolo a tres personas, Alice, Bob y Carlos, suponemos que tienen como claves secretas propias $a,b$  y $c$. La idea sería crear algo similar al protocolo con dos personas, por lo que los mensajes irán enviados de una manera cíclica, es decir, contemplando las cadenas Alice\-Bob\-Carlos, Bob\-Carlos\-Alice y Carlos\-Alice\-Bob.

\- Alice calcula $A \equiv g^a \mod p$  y lo envía a Bob. Bob calcula $A_2 = A^b \equiv g^{ab}   \mod   p$  y lo envía a Carlos, que calcula $ g^{abc}   \mod   p$.

\- Bob calcula $B\equiv g^b   \mod p$ y lo envía a Carlos. Carlos calcula $B_2 = B^c \equiv g^{bc}   \mod   p$  y lo envía a Alice, que calcula $g^{abc}   \mod   p$ .

\- Carlos calcula $C\equiv g^c   \mod p$  y se lo envía a Alice, que calcula $C_2 = C^a \equiv g^{ac}   \mod   p$  y lo envía a Bob, que termina con $ g^{abc}   \mod   p$.

De esta manera, la clave común es $g^{abc} \mod   p$ .


**17. Sea p=971 y g=6. La clave pública en el criptosistema de ElGamal es (p,q,424). Cifra GAUSS usando esta clave pública y describe como Alice puede descifrar el mensaje.**

Tomamos un alfabeto de 27 letras donde el 0 es el espacio y pasamos a números el mensaje a cifrar
\begin{equation*}
m \equiv 7 \cdot 27^4+1 \cdot 27^3 + 21\cdot 27^2 +19 \cdot 27 +19 \equiv 3755611 \equiv 754 \ mod \ 971
\end{equation*}

Tomamos una b que solo sabremos nosotros y escogemos por ejemplo b=6. Ahora para cifrar envíamos $(B,C) \equiv_p (g^b,A^bm)$
\begin{equation*}
B \equiv g^b \equiv 6^6 \equiv 46656 \equiv 48 \ mod \ 971
\end{equation*}
\begin{equation*}
C \equiv A^bm \equiv 424^6 \cdot 754 \equiv 500 \ mod \ 971
\end{equation*}
Por lo tanto, el mensaje cifrado sería (48,500), en realidad el mensaje cifrado es 500 pero sin el 48 Alice no podría descifrarlo.<br>
Ahora, para desencriptar el mensaje Alice debería tener un a secreto que solo ella sabe. Primero calcularía, $B^a$, después calcularía el inverso de dicho número módulo 971 ($B^{-a}$), el cual existe por ser 971 primo. Una vez ya tenemos $B^{-a}$ para descifrar solo tendremos que multiplicar dicho número  por el mensaje cifrado C y obtenemos m:
\begin{equation*}
B^{-a}C \equiv g^{-ba}A^bm \equiv g^{-ab}g^{ab} m \equiv m \ mod \ 971
\end{equation*}



18. **Diseñe un criptosistema de Merkle\-Hellman para cifrar 2 letras de cada vez y cifra FIM.**



Usamos un alfabeto de 27 letras. Para cifrar 2 letras de cada vez, hacemos lo siguiente: dadas dos letras $x,y\in \{0,1,\dots,26\}$, el par de letras se corresponderá con el entero $27x+y \in \{0,1,\dots,728\}$. Entonces $k=\lfloor \log_2(728) \rfloor +1 = 9+1 = 10$. Así, por ejemplo tendremos:

$\_$ $\_ =27\cdot 0 +0 = 0 = (0000000000)_2$, $A \_ = 27\cdot 1 + 0 = 27 = (0000011011)_2$, $AA = 27\cdot 1 +1 = 28 = (0000011100)_2$, $\dots$, $ZZ = 27\cdot 26 +26 = 728 = (1011011000)_2$

Ahora construimos una secuencia supercreciente:

$(v_0, \dots, v_{k-1}) = (2,3,7,15,31,60,119,238,477,955)$

Tomamos $n > v_0 + \dots + v_{k-1} = 1907$, por ejemplo $n = 1910$, y un entero $a$, $0<a<n$ tal que $(a,n)=1$, por ejemplo $a=9$.

Calculamos $b \equiv a^{-1}$ mod $n \equiv 9^{-1}$ mod 1910 \equiv 849$

$w_0 \equiv av_0$ mod $n \equiv 9 \cdot 2$ mod $1910 \equiv 18$

$w_1 \equiv av_1$ mod $n \equiv 9 \cdot 3$ mod $1910 \equiv 27$

$w_2 \equiv av_2$ mod $n \equiv 9 \cdot 7$ mod $1910 \equiv 63$

$w_3 \equiv av_3$ mod $n \equiv 9 \cdot 15$ mod $1910 \equiv 135$

$w_4 \equiv av_4$ mod $n \equiv 9 \cdot 31$ mod $1910 \equiv 279$

$w_5 \equiv av_5$ mod $n \equiv 9 \cdot 60$ mod $1910 \equiv 540$

$w_6 \equiv av_6$ mod $n \equiv 9 \cdot 119$ mod $1910 \equiv 1071$

$w_7 \equiv av_7$ mod $n \equiv 9 \cdot 238$ mod $1910 \equiv 232$

$w_8 \equiv av_8$ mod $n \equiv 9 \cdot 477$ mod $1910 \equiv 473$

$w_9 \equiv av_9$ mod $n \equiv 9 \cdot 955$ mod $1910 \equiv 955$

La clave pública para cifrar es:

$\{w_0, \dots, w_{k-1}\} = \{18,27,63,135,279,540,1071,232,473,955\}$

Queremos cifrar FIM, tomando las letras de 2 en 2:

FI $ = 27\cdot 6 + 9 = 171 = (0010101011)_2$

M\_ $= 27\cdot 13 + 0 = 351 = (0101011111))_2$

Ciframos FI como: $18+27+135+540+232 = 952$

M\_ como: $18+27+63+135+279+1071+473 = 2066$

19. **Considera la codificación habitual en bits del alfabeto de 27 letras \(**$A = (00001)_2, B = (00010)_2$ ...**\). Considera el cirptosistema de Merckle\-Hellman y la llave** $((v_0, ..., v_{k-1}), n, a) = ((2,3,7,15,31), 61, 17)$. **Cifra, letra a letra, el mensaje "SIM"** 



Notemos que la secuencia de $v_i$  dada es supercreciente. Calculamos la clave pública,{ ${ w_i = av_i, i=0,1,2,3,4 } $ }, que usaremos para cifrar letra a letra. 


In [3]:
def w_i(secuencia, a, n):
    w_i = []
    for i in secuencia:
        w_i.append(pow(a*i,1,n))
    return w_i

secuencia = [2,3,7,15,31]
print('La secuencia de w_i es :', w_i(secuencia,17,61))

La secuencia de w_i es : [34, 51, 58, 11, 39]


Ahora, para el cifrado, obtengamos la representación binaria de cada una de las letras del mensaje a cifrar. Recordemos que estamos con 27 letras en el alfabeto\(contamos la ñ\):

- $S \equiv 19 \equiv (10011)_2$ 
- $I \equiv 9 \equiv (01001)_2$ 
- $M \equiv 13 \equiv (01101)_2$ 



La forma de codificar es hacer una asociación entre la representación binaria y la secuencia de $w_i$.

- $S \equiv 19 \equiv (10011)_2$  $\rightarrow$ Encriptamos como $34+51+39 =124$ 
- $I \equiv 9 \equiv (01001)_2$ $\rightarrow$ Encriptamos como $34+11 = 45$ 
- $M \equiv 13 \equiv (01101)_2$ $\rightarrow$ Encriptamos como $34+58+11 = 103$ 



De esta forma, el mensaje 'SIM' enviado letra a letra sería $(124, 45, 103)$.


**20. Por el algoritmo de Tonelli-Shanks encuentra las raices de $a = 183764 \ mod \ p = 2234261$. (Nota que $p − 1 = 2^2\cdot 558565$).**

Para aplicar el algoritmo de Tonelli-Shanks tenemos a=183764, p=2234261 y después por la forma de p-1 tenemos $d=558565$ y $s=2$. Ahora calculamos $r\equiv a^{\frac{d+1}{2}} \equiv a^{279283} \equiv 487463 \ mod \ p$ y $t \equiv a^d \equiv 183764^{558565}\equiv 2234260 \equiv -1\ mod \ p $. Como t es distinto de 1 tenemos que seguir con el algoritmo, para ello tenemos que buscar primero un no residuo cuadrático de p. Como p es de la forma 8k+5, sabemos que 2 no es residuo cuadrático de p, por tanto tomamos z=2. Ahora calculamos $c_1\equiv z^{d}\equiv 2^{558565} \equiv 901202$, calculamos $i_1$ como el menor valor $1 \leq i_1 < s$ tal que $t^{2^{i}} \equiv 1$ y claramente viendo que s=2, $i_1$ tiene que ser 1, ademas como $t \equiv -1$ vemos que $t^{2^1} \equiv 1$. Ahora tenemos que $b_1 \equiv c^{2^{s-1-i}} \equiv c^{2^0} \equiv c \equiv 901202$. Calucalamos los nuevos r y t haciendo $r_2 \equiv r_1 \cdot b_1 \equiv 487463 \cdot 901202 \equiv 2232706 \ mod \ p$ y $t_2 \equiv t_1 \cdot b_1^2 \equiv (-1) \cdot (-1) \equiv 1 \ mod \ p$. Por tanto sabemos que $r_2 \equiv 2232706$ es una raíz de a=183764. Lo comprobamos y vemos como $2232706^2 \equiv 183764 \ mod \ 2234261$. (Como a es disitnto de 0 sabemos que tendrá otra raíz la cual será 1555 ya que $2232706 \equiv -1555 \ mod \ 2234261$).

21. **Hacer los cálculos para encriptar y desencriptar el mensaje con ElGamal siendo** $m = 7$**,** $p = 97$**,** $g = 5$ **y la clave secreta de Alice** $a = 58$**. Considera que Bob elije** $b = 17$**.**



$p=97$, $g=5$ raíz primitiva de $p$. 

Alice elige un exponente secreto $a=58$ y manda a Bob $A \equiv g^a$ mod $p \equiv 5^{58}$ mod $97 \equiv 44$

Bob elige un exponente $b=17$ (que mantiene en secreto) y para enviar $m=7$ lo cifra y envía a Alice $(g^b$ mod $p, A^b \cdot m) = (5^{17}$ mod $97, 44^{17} \cdot 7) = (83,67)$.

Alice recibe $(B,c) = (83,67)$ y para recuperar el mensaje $m$ calcula $B^{-a} = 83^{-58} = (83^{58})^{-1} \equiv 65^{-1}$ mod $p\equiv 3$ y finalmente $m \equiv B^{-a} \cdot c$ mod $p \equiv 3 \cdot 67$ mod $97 \equiv 7$.


**23.  Imagina que Miguel tiene una amante, Marta, con la que quiere intercambiar mensajes secretos usando el sistema ElGamal. Imagina que Miguel usa el mismo exponente secreto para mandar dos mensajes $M_1$ y $M_2$ a Marta. ¿Qué sucede si su novia Olga intercepta los mensajes? ¿Los puede leer? ¿Qué pasa si se imagina lo que hay en uno de ellos? ¿Consigue leer el otro mensaje?**


Supongamos que Miguel envía dos mensajes $m$ y $\bar{m}$ con el mismo b, entonces el oyente obtendrá
\begin{equation*}
c \equiv A^b m \ mod \ p \ y \ \bar{c} \equiv A^b \bar{m} \ mod \ p
\end{equation*}

De aquí podemos sacar por tanto que $\bar{c} c^{-1} \equiv \bar{m} m^{-1} \ mod \ p$. Es decir que si Olga no conoce ninguno de los mensajes no va a poder descifrarlos, pero si Olga se imagina lo que pone en uno de los dos mensajes entonces si podrá leer el otro mensaje, ya que supongamos que conoce el contenido de m, entonces
\begin{equation*}
\bar{m} \equiv \bar{c} c^{-1} m \ mod \ p
\end{equation*}
y como conocemos todos los términos de la derecha podemos calcular $\bar{m}$, es decir, descifrar el otro mensaje.



In [20]:
from math import trunc,log
def num_posiciones(num, len_alfabeto):
    k = trunc(log(num,len_alfabeto))
    posiciones =  []
    contador = k
    for i in range(0,k):
        coc = num//(len_alfabeto**(contador))
        resto = num%(len_alfabeto**(contador))
        num = resto
        contador -= 1
        posiciones.append(coc)
    posiciones.append(resto)
    return posiciones

print(num_posiciones(441,27))
    
    
        
    

[16, 9]
